buffer/src/main/java/io/netty/buffer/ByteBufUtil.java

```diff
@@ -706,6 +706,66 @@ public static void copy(AsciiString src, int srcIdx, ByteBuf dst, int length) {
     }
 
     /**
+     * Creates a retained duplicate of the specified {@link ByteBuf}.
+     */
+    public static ByteBuf rduplicate(ByteBuf buf) {
+        final int readerIndex = buf.readerIndex();
+        final int writerIndex = buf.writerIndex();
+
+        final AbstractByteBuf duplicate;
+        if (buf instanceof SlicedAbstractByteBuf) {
+            final SlicedAbstractByteBuf oldSlice = (SlicedAbstractByteBuf) buf;
+            duplicate = PooledSlicedByteBuf.newInstance(oldSlice.unwrap(), 0, buf.capacity(), oldSlice.adjustment());
+            duplicate.setIndex(readerIndex, writerIndex);
+            duplicate.markReaderIndex();
+            duplicate.markWriterIndex();
+        } else if (buf instanceof DuplicatedAbstractByteBuf) {
+            duplicate = PooledDuplicatedByteBuf.newInstance((AbstractByteBuf) buf.unwrap(), readerIndex, writerIndex);
+        } else if (buf instanceof AbstractByteBuf) {
+            duplicate = PooledDuplicatedByteBuf.newInstance((AbstractByteBuf) buf, readerIndex, writerIndex);
+        } else if (buf instanceof SlicedByteBuf) {
+            duplicate = new SlicedByteBuf(buf, 0, buf.capacity());
+            duplicate.setIndex(readerIndex, writerIndex);
+            duplicate.markReaderIndex();
+            duplicate.markWriterIndex();
+            duplicate.retain();
+        } else {
+            duplicate = new DuplicatedByteBuf(buf, readerIndex, writerIndex);
+            duplicate.retain();
+        }
+
+        return duplicate;
+    }
+
+    /**
+     * Creates a retained slice of the specified {@link ByteBuf}.
+     */
+    public static ByteBuf rslice(ByteBuf buf) {
+        final int index = buf.readerIndex();
+        return rslice(buf, index, buf.writerIndex() - index);
+    }
+
+    /**
+     * Creates a retained slice of the specified {@link ByteBuf}.
+     */
+    public static ByteBuf rslice(ByteBuf buf, int index, int length) {
```

Comment:
No, they can be used by a user who wants to implement a `ByteBuf` without extending `AbstractByteBuf`.